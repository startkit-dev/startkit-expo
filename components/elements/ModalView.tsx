import { useNavigation } from "@react-navigation/native"
import { useSafeAreaInsets } from "react-native-safe-area-context"

import { useRootBackgroundColor } from "@/hooks/use-root-background-color"

import { Icon } from "./Icon"
import { Pressable } from "./Pressable"
import { View, type ViewProps } from "./View"

type ModalViewProps = ViewProps & {
  /**
   * A custom root-level view background color for this modal, passed to
   * useRootBackgroundColor.
   */
  backgroundColor?: string
}

export function ModalView({
  backgroundColor,
  children,
  ...props
}: ModalViewProps) {
  useRootBackgroundColor(backgroundColor)
  const { top } = useSafeAreaInsets()
  const navigation = useNavigation()

  const onBackPressed = () => {
    navigation.goBack()
  }

  return (
    <View className="flex-1 justify-start ios:items-center">
      {/* Swipe bar for iOS */}
      <View className="z-10 mx-auto my-4 h-1 w-12 rounded-xl bg-slate-900 opacity-70 android:hidden dark:bg-slate-50" />
      {/* Back arrow for Android */}
      <Pressable
        haptics
        className="h-10 w-10 items-center justify-center active:opacity-50 ios:hidden"
        style={{ marginTop: top }}
        onPress={onBackPressed}
      >
        <Icon name="x" size={24} />
      </Pressable>
      <View {...props} className="w-full flex-1">
        {children}
      </View>
    </View>
  )
}
