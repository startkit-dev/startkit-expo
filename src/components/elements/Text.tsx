import { clsx } from "clsx"
import {
  Text as DefaultText,
  type TextProps as DefaultTextProps
} from "react-native"

export type TextProps = DefaultTextProps & {
  /**
   * Toggle to use the default monospace font
   */
  mono?: boolean
}

/**
 * A Wrapper round the <Text> component which applies our basic styles,
 * considering dark mode, etc.
 */
export function Text({ mono, ...props }: TextProps) {
  return (
    <DefaultText
      className={clsx(
        "text-black dark:text-white",
        mono && "font-mono font-normal"
      )}
      {...props}
    />
  )
}
