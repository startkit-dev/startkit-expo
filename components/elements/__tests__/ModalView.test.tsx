import { create } from "react-test-renderer"

import { ModalView } from "../ModalView"
import { Text } from "../Text"

test("renders correctly", () => {
  const tree = create(
    <ModalView>
      <Text>ModalView test</Text>
    </ModalView>
  ).toJSON()

  expect(tree).toMatchSnapshot()
})
