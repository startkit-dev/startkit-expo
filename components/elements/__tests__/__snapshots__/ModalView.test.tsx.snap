// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <View />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "marginTop": 0,
      }
    }
  >
    <Text />
  </View>
  <View>
    <Text>
      ModalView test
    </Text>
  </View>
</View>
`;
